File: src/gamestate/GameStateMachine.java
package gamestate;

import handlers.controls.Controls;
import main.Main;
import org.newdawn.slick.GameContainer;
import org.newdawn.slick.Graphics;
import region.RegionManager;

import java.util.ArrayList;

public class GameStateMachine {
	public static final int EFFECT_GRASS_A = 0;
	
	public static final int EFFECT_GRASS_B = 1;
	
	public static final int EFFECT_WATER_A = 2;
	
	public static final int EFFECT_WATER_B = 3;
	
	public static final int EFFECT_CAVE_A = 4;
	
	public static final int EFFECT_CAVE_B = 5;
	
	public static final int inGameState = 0;
	
	public static final int battleState = 1;
	
	private static ArrayList<AbstractGameState> gameStates = new ArrayList<>();
	
	private static int currentState;
	
	private static long transitionStartTicks = 0;
	
	private static boolean battleTransitioning = false;
	
	private static int transitionEffect;
	
	public static void update(GameContainer gc) {
		gameStates.get(currentState).update(gc);
		Controls.INSTANCE.update(gc);
		if (battleTransitioning) {
			if (Main.ticks >= transitionStartTicks + 32 + 128) {
				//Done transitioning
				battleTransitioning = false; //TODO(" Maybe use more bools to show which part of the transition we're in")
			} else if (Main.ticks >= transitionStartTicks + 128) {
				//TODO(" Transition In")
			} else if (Main.ticks >= transitionStartTicks + 64) {//TODO(" Set proper transition effect")
				//Transition Out
				setGameState(battleState);
			}
		}
	}
	
	public static void render(GameContainer gc, Graphics g) {
		gameStates.get(currentState).render(gc, g);
		if (battleTransitioning) {
			//TODO(" battle transition")
			//draw transition image
			//something like "TransitionAnimation" + transitionEffect + ".gif"
			//using transitionStartTicks - Main.ticks to get the animation frame
		}
	}
	
	public static void init(GameContainer gc) {
		gameStates.add(new InGameState()); //0
		gameStates.add(new BattleState()); //1
		currentState = inGameState;
		RegionManager.init(gc);
	}
	
	public static AbstractGameState getGameState() {
		return gameStates.get(currentState);
	}
	
	public static void setGameState(int state) {
		currentState = state;
	}
	
	public static AbstractGameState getGameState(int stateId) {
		return gameStates.get(stateId);
	}
	
	public static void transitionToBattle(int transition) {
		battleTransitioning = true;
		transitionEffect = transition;
		transitionStartTicks = Main.ticks;
	}
	
	public static boolean isTransitioning() {
		return battleTransitioning;
	}
}

File: src/gamestate/GameStateMachine.kt
package gamestate

import gui.GuiManager
import util.controls.Controls
import main.Main
import org.newdawn.slick.GameContainer
import org.newdawn.slick.Graphics
import region.RegionManager
import util.Resources

import java.util.ArrayList

object GameStateMachine {
    const val EFFECT_GRASS_A = 0
    const val EFFECT_GRASS_B = 1
    const val EFFECT_WATER_A = 2
    const val EFFECT_WATER_B = 3
    const val EFFECT_CAVE_A = 4
    const val EFFECT_CAVE_B = 5

    private const val inGameState = 0
    private const val battleState = 1

    private val gameStates = ArrayList<AbstractGameState>()
    private var currentState: Int = 0

    private var transitionStartTicks: Long = 0
    private var isTransitioning = false
    private var transitionEffect: Int = 0

    val gameState: AbstractGameState
        get() = gameStates[currentState]

    fun update(gc: GameContainer) {
        Resources.update()
        Controls.update(gc)

        gameStates[currentState].update(gc)


        //TODO("Move transitions elsewhere | Complete transitions")
        if (isTransitioning) {
            when {
                Main.ticks >= transitionStartTicks + 32 + 128 -> //Done transitioning
                    isTransitioning = false
                Main.ticks >= transitionStartTicks + 128 -> {
                    //Transition In
                }
                Main.ticks >= transitionStartTicks + 64 -> //Transition Out
                    setGameState(battleState)
            }
        }

        GuiManager.update(gc)
    }

    fun render(gc: GameContainer, g: Graphics) {
        gameStates[currentState].render(gc, g)
        GuiManager.render(gc, g)

        if (isTransitioning) {
            //TODO(" battle transition")
            //draw transition image
            //something like "TransitionAnimation" + transitionEffect + ".gif"
            //using transitionStartTicks - Main.ticks to get the animation frame
        }
    }

    fun init(gc: GameContainer) {
        gameStates.add(InGameState) //0
        gameStates.add(BattleState) //1
        currentState = inGameState
        RegionManager.init(gc)
    }

    private fun setGameState(state: Int) {
        currentState = state
    }

    fun getGameState(stateId: Int): AbstractGameState {
        return gameStates[stateId]
    }

    fun transitionToBattle(transition: Int) {
        isTransitioning = true
        transitionEffect = transition
        transitionStartTicks = Main.ticks
    }
}
